File: 2021/day01.hs
import           Data.List (tails)

main :: IO ()
main = do
  input <- readFile "./01/input.txt"
  let list = read <$> lines input
  print $ "Part 1: " <> show (solve1 list)
  print $ "Part 2: " <> show (solve2 list)

solve1 :: [Int] -> Int
solve1 = length . filter (uncurry (<)) . pairs

solve2 :: [Int] -> Int
solve2 input = length . filter (uncurry (<)) $ pairs (map sum . windows 3 $ input)

pairs :: [a] -> [(a,a)]
pairs xs@(_:xs') = zip xs xs'

windows :: Int -> [a] -> [[a]]
windows m = foldr (zipWith (:)) (repeat []) . take m . tails

File: 2021/day01.rb
input = File.readlines("01/input.txt").map(&:to_i)

class Array
  def window(size)
    return [] unless size > 0 && size <= self.length
    Enumerator.new do |y|
      first, last = 0, size-1
      while first <= self.length - size do
        y << self[first..last]
        first, last = first+1, last+1
      end
    end
  end
end

def solve1(input)
  input.window(2)
    .filter { |a, b| a < b }
    .length
end

def solve2(input)
  input.window(3).map(&:sum).window(2)
    .filter { |a, b| a < b }
    .length
end

puts "Part 1: #{solve1(input)}"
puts "Part 2: #{solve2(input)}"
